### PR TITLE
Convert Trainer doc page to MarkDown

### LINKED ISSUE #14730: Fix docs pointers for DeepSpeed
In this [section](https://huggingface.co/docs/transformers/master/en/main_classes/trainer#deepspeed), the reader is pointed to several links that are broken:

* https://huggingface.co/docs/transformers/master/en/main_classes/trainer#deepspeed-trainer-integration
* https://huggingface.co/docs/transformers/master/en/main_classes/trainer#deepspeed-installation
* https://huggingface.co/docs/transformers/master/en/main_classes/trainer#deepspeed-multi-gpu
* https://huggingface.co/docs/transformers/master/en/main_classes/trainer#deepspeed-one-gpu
*https://huggingface.co/docs/transformers/master/en/main_classes/trainer#deepspeed-config-passing
* etc

All of them should be changed like this:

`https://huggingface.co/docs/transformers/master/en/main_classes/trainer#(.*)` --> `https://huggingface.co/docs/transformers/master/en/main_classes/deepspeed#(.*)`

I would do it myself, but I see that the code uses e.g. `:ref:`deepspeed-notebook` and not sure how to point to a section in another page 🤷‍♂️  It looks like a trivial change though

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

The `doc-builder` only resolves `ref` when they point to sections in the same page, so this PR fixes #14730 by converting the doc page to MarkDown and fixing all links. In passing, it removes mention of the `TFTrainer` which is now deprecated.

### CODE REVIEW

File: docs/source/main_classes/trainer.mdx [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+- Using `--sharded_ddp zero_dp_3` requires wrapping each layer of the model in the special container
+  `FullyShardedDataParallelism` of fairscale. It should be used with the option `auto_wrap` if you are not
+  doing this yourself: `--sharded_ddp "zero_dp_3 auto_wrap"`.
+
+
+### DeepSpeed
+
+
+Moved to [Trainer DeepSpeed integration](deepspeed#trainer-deepspeed-integration).
+
```
Comment (sgugger) [inline docs/source/main_classes/trainer.mdx:449]: I kept everything as is @stas00 but maybe it's time to remove everything below and only leave this first line? The rendering on the doc is not super nice, as seen [here](https://huggingface.co/docs/transformers/main_classes/trainer#deepspeed)

Comment (stas00) [reply docs/source/main_classes/trainer.mdx:449]: The intention was to preserve old links. so if someone on another site/forum linked to a specific section the user following this link will be able to find the new location after the move.

If you remove it, the anchor will fail and the user will get the top of the page and won't even know about that main: `moved to ...` 

One easy way to fix this on the web server level is to create a redirect file which matches the old link with the anchor and redirects it to a new location, .e.g;

```
https://huggingface.co/docs/transformers/main_classes/trainer#deepspeed| https://huggingface.co/docs/transformers/master/en/main_classes/trainer#deepspeed-trainer-integration
... same for all other redirects ...
```

we already have a redirect rule from `.html` to non-html link, so these need to be added there.

It appears hf.co is running on `nginx` webserver, so need to look up the format for redirects on that server.

then we can nuke those "poor man's redirects" in the doc itself.

Comment (stas00) [reply docs/source/main_classes/trainer.mdx:449]: should we discuss this on #doc-frontend?

Comment (sgugger) [reply docs/source/main_classes/trainer.mdx:449]: Sure!

Comment (julien-c) [reply docs/source/main_classes/trainer.mdx:449]: nginx does not see anchor links, this is a purely client-side element

File: .github/workflows/build_doc_test.yml [hunk 2]
```diff
@@ -46,4 +46,4 @@ jobs:
 
       - name: Make documentation
         run: |
-          doc-builder build transformers ./transformers/docs/source
+          doc-builder build transformers ./docs/source
```
Comment (sgugger) [inline .github/workflows/build_doc_test.yml:49]: An error actually made me realize the path was wrong all along 🤦

File: docs/source/main_classes/trainer.mdx (hunk 1 above)
Comment (stas00) [reply docs/source/main_classes/trainer.mdx:449]: Thank you, @julien-c, for sharing that. Indeed the browser doesn't send the anchor part to the server :(

That means we can't do the redirect in the rules file :(

It's your call Sylvain, if you prefer clean and neat over supporting existing links then please remove all but the first one.

Alternatively, we could discuss and agree on a general policy - when we rename headers or move content to a different file, do we support old links or not and at what cost. And depending on the outcome of that discussion then this decision about the current question can be made.

Comment (stas00) [reply docs/source/main_classes/trainer.mdx:449]: > In this case, how about we just leave the HTML anchors and remove all visible content? A user that tries a "dead" link will then see "Moved to [Trainer DeepSpeed integration]" and can then easily find their way from there.

If it works that would be great. Could you do a single link mockup to see what you mean exactly?

Something like:
```
<h3><a href="#deepspeed"></a>DeepSpeed</h3>
<p>Moved to <a href="/docs/transformers/master/en/main_classes/trainer#deepspeed-trainer-integration">deepspeed-trainer-integration</a>.</p>
```

but add it to a non-visible class?

**Review by LysandreJik — APPROVED** [review]
Thanks for showing the way with the conversion!

File: docs/source/main_classes/trainer.mdx (hunk 1 above)
Comment (sgugger) [reply docs/source/main_classes/trainer.mdx:449]: I'll open a fresh PR to discuss this and merge this one in the meantime to fix the related issue :-)

**Pull request merged**